source/extensions/stat_sinks/metrics_service/grpc_metrics_service_impl.h

```diff
@@ -74,24 +77,111 @@ class GrpcMetricsStreamerImpl : public Singleton::Instance, public GrpcMetricsSt
 using GrpcMetricsStreamerImplPtr = std::unique_ptr<GrpcMetricsStreamerImpl>;
 
 /**
- * Stat Sink implementation of Metrics Service.
+ * Stat Sink that flushes metrics via a gRPC service.
  */
-class MetricsServiceSink : public Stats::Sink {
+template <class ResponseProto> class MetricsServiceSink : public Stats::Sink {
 public:
   // MetricsService::Sink
-  MetricsServiceSink(const GrpcMetricsStreamerSharedPtr& grpc_metrics_streamer,
-                     const bool report_counters_as_deltas);
-  void flush(Stats::MetricSnapshot& snapshot) override;
+  MetricsServiceSink(const GrpcMetricsStreamerSharedPtr<ResponseProto>& grpc_metrics_streamer,
+                     const bool report_counters_as_deltas)
+      : grpc_metrics_streamer_(grpc_metrics_streamer),
+        report_counters_as_deltas_(report_counters_as_deltas) {}
+  void flush(Stats::MetricSnapshot& snapshot) override {
+    metrics_.Clear();
```

Comment:
Would it make sense to split off all the code that deals with populating `metrics_` into another class (composite or base class) so we can avoid blowing up the header file with implementation details?  Basically cut down on what code needs to be templated

Reply:
@snowp I debated back and forth on this one. I settled on the templated class because that meant that:
a) in envoy mobile I would not even need to define a derived class, the config factory just templates this class (in the case of making this a base class)
b) I don't need to pass config params -- like the counters as deltas bool -- to class functions (in the case of a composite class).

But I am more than happy to make a move to either if you think it is preferable. Do you have a preference on either (base or composite class)?

Reply:
I think with my suggestion you'd still be using a templated class, it's just that all the code that doesn't really care about the template params are moved into another class. This would mostly just be about working around the fact that C++ wants you to put the implementation details in headers for templated class: you'd have a thin templated class that interfaces with a non-templated class that doesn't require you to put all the details in the header file.

If this proves difficult it's probably fine to leave it as is, I don't feel that strongly about it

Reply:
Yeah I am on the same page as you, I understand the c++ need of having implementation in headers for templated classes. 

I was saying that the current layout prevents having the thin, templated sink class, and thus prevents having to have one in envoy mobile as well.

Not difficult in any of the different layouts, just a preference.

Reply:
nvm, I see the setup you _actually_ meant. Will do